### DISCUSSION TITLE
关于Milvus性能限制的相关问题

### DISCUSSION THREAD

**Discussion by xingfan-git** [milvus-io/milvus#22264] [Q&A and General discussion] [answered]
您好，我们现在正在使用Milvus，在使用过程中遇到了一些问题，希望您可以帮忙解答一下

1. 在文档中，我们看到connection limit是65536；但是在FAQ中又看到“Python SDKs for Milvus v0.9.0 or higher have a connection pool. The number of connections in a connection pool has no upper limit.” 请问connection现在是否还有数量限制呢？以及这个限制具体是指什么，是一台机器上connection（context）的最大数量，还是一个Milvus server可以连接的最大用户数量呢？
2. SDK的collection.load()并没有返回值。那么我们有什么办法可以拿到一个collection的load状态呢（比如unload/loaded/loading）？
3. 对于正在loading的collection，如果发送search请求，会直接报错，还是持续等待呢？
4. **我们用了一个9,990,000个96维vector，size应该是～3.6G，但是在load之后我们发现机器的memory用掉将近40G，可以请问一下这个是正常的嘛？为什么会比dataset本身超出这么多呢？Milvus对于collection load之后的内存管理有什么优化的方法嘛？**
<img width="1568" alt="image" src="https://user-images.githubusercontent.com/50465320/219600878-7c002afd-d445-45a9-90b6-2964a5d6d2db.png">

**Comment by yhmo** [comment, marked as answer]
1. 在早期版本(<= 1.1) 的python sdk中确实是做了一个连接池，并且该连接池的数量没有上限。在目前2.x版本的python sdk中，我并没有看到有连接数的限制，文档中的65536我要去核实一下。这里的连接指的是milvus client和server只间的连接，milvus使用grpc来传输，grpc理论上没有连接数的限制(我也是看stackoverflow说的，不对的话帮我纠正)。
2. 在最新的2.2.3版本中，对应的最新的python sdk和java sdk都增加了一个叫getLoadState的接口，可返回load状态。之前的版本是没有这个接口的
3. 还没loading完成的，search都会报错
4. 这个好像不太正常，你的milvus是不是standalone的？如果是standalone，那么querynode和indexnode都是在同一个机器上，同时那些第三方服务比如minio，etcd也在同一个机器上，他们的内存使用量加起来会比3.6G多，但也不至于多10倍。正常的话多个两三倍吧。具体为什么用这么多，得有更多的信息。

**Reply by xingfan-git** [reply]
首先感谢您的回答，这里有一些补充：

1. connection limitation是在这里看到的：https://milvus.io/docs/limitations.md#Number-of-resources
<img width="907" alt="image" src="https://user-images.githubusercontent.com/50465320/219655918-a5f0d879-9404-4b54-8edc-0b2de91f6da1.png">

2. 谢谢。我从roadmap看到2.2.3原计划2月8日发布，目前看起来像是延时了，请问你知道目前的计划发布日期吗？

3. 谢谢

4. 我们不是standalone的，是在一个aks集群上部署的，一共有四台机器（16核，128G内存），对应了图中的四条线。queryNode、dataNode和indexNode的数量都是4。图中最高的那条线对应的机器有3个queryNode，1个indexNode和1个dataNode，同时有一个minIO，具体如下1图。想请问一下load时候是每一个queryNode都需要load一整份collection嘛，还是会把collection的segment均分到所有的queryNode上（我理解是第二种情况？）？不知道这些信息会不会对您分析问题有帮助：

（内存占用最高的机器的pods）
<img width="894" alt="image" src="https://user-images.githubusercontent.com/50465320/219658804-0d6b8614-1e7c-4d6b-97df-83c4e3bc3faa.png">

（整个cluster的pods）
<img width="835" alt="image" src="https://user-images.githubusercontent.com/50465320/219659598-7e77df26-2fb7-40e2-a422-a9a0aed2a068.png">


再一次谢谢您！

**Reply by yhmo** [reply]
- MIlvus 2.2.3已经在2月10号发布了： https://github.com/milvus-io/milvus/releases/tag/v2.2.3，与之对应的最新的Java sdk版本也是2.2.3，最新的python sdk版本是2.2.2

- load是把collection的segment尽量均分到各个querynode上去。那台机器有3个queryNode，1个indexNode和1个dataNode，同时有一个minIO，那么内存的消耗有这几个方面：
   a. querynode把segment加载到内存，如果你调用load()时设置了replica_number > 1，每个segment在内存里有多份copy，那么内存的使用量就是按倍数计算
   b. 如果indexnode正在构建索引，那么构建索引期间也是消耗内存的，一般使用量跟单个segment的大小相关
   c. 如果有增量数据输入，datanode会有缓存，直到调用flush()或者内部触发flush，才会把缓存写入存储系统。如果大量collection或者大量partition，每个partition都插入一点数据，数据量又没达到触发flush缓存的量，那么这些缓存加起来就会占用很可观的内存
   d. minio这些第三方组件本身也会占用一些内存，但不多
一般来说，如果场景里带有a和c，那么内存占用量是会很大。

**Reply by xingfan-git** [reply]
- 请问一下python sdk的2.2.3的版本大概什么时候会发布呢
- 我们在insert之后先做了flush，然后才create_index，最后在load时候没有设置任何参数，所以应该没有你提到的a或是c的情况。我们去查了一下各个container的内存占用，进一步发现了一下比较奇怪的事情，想麻烦您帮忙分析一下：
a. 首先我想你们的float vector中的float应该是32位的（如果不对，请指正），那么下面这个数据集（9,990,000条数据）大小应该是~3.65G
![image](https://user-images.githubusercontent.com/50465320/220009340-30956f89-ff1b-4d11-949b-a867fcec59a3.png)
在四个queryNode中，数据大小一共为14.3G（如下图），接近为原数据的4倍。按照您之前的描述，这个应该是不合理的，请问这个一般是由什么导致的呢？
![image](https://user-images.githubusercontent.com/50465320/220009900-44f0f4da-d451-4aca-bde0-edeb5f2b4280.png)
b. 我们发现有一些pulsar相关的container，pulsar-bookie、pulsar-broker和pulsar-proxy，一共占用了14.25G的内存（静态，没有request）。这个pulsar-bookie看上去只是一些消息队列，不是很清楚为什么会占用这么大的内存，请问您有什么建议嘛？
![image](https://user-images.githubusercontent.com/50465320/220012598-1ac470a4-a428-46a8-a410-16b5a2c177fa.png)
c. 我们发现静态时，indexNode和dataNode（各四个），一共也会占用将近8G内存，请问这个是正常的嘛？
![image](https://user-images.githubusercontent.com/50465320/220013282-da5f95e3-d4bd-46a6-a1d7-5de5d040be14.png)
![image](https://user-images.githubusercontent.com/50465320/220013322-ef06308e-33d5-4266-8f9c-ab243a61937c.png)

**Reply by xiaofan-luan** [reply]
Milvus 用Go做的，Pulsar是用Java做的，GC语言不会回收内存释放给操作系统，这部分内存并非被使用了，只是没有释放给操作系统。正常我们对于Milvus的推荐大小就是单台机器不小于8G。如果1000w数据，一个DataNode和一个indexNode就可以，pulsar也可以适当调小内存占用

**Reply by xingfan-git** [reply]
好的，谢谢，关于pulsar这部分的理解了，但是对于其他部分的问题，还是麻烦您解答一下：
- queryNode与origin dataset的size相差四倍，这样子是正常的嘛？有没有什么办法可以减少内存使用呢？如果一直按照四倍线性增长的话，可能会超出我们对memory的预期，而且似乎和yihua前面讲的2~3倍有一些比较大的出入；
- python sdk 2.2.3似乎还没有发布，想问一下有一个大概的预估发布时间嘛？如果没有的话，请问你们有公开getLoadStatus相关的Restful API给用户嘛？
- 有另一个关于并发的问题：假设client 1 release了一个collection，并且开始建index，此时client 2要进行查询，load collection，这时候client 2的操作会失败嘛？如果会的话，client 2有什么办法可以知道现在collection有正在进行的建index的操作嘛（比如有暴露接口可以查询正在进行中的任务，或者是检查是否满足load collection的条件嘛）？

提前谢谢您的解答！

**Reply by yhmo** [reply]
> 好的，谢谢，关于pulsar这部分的理解了，但是对于其他部分的问题，还是麻烦您解答一下：
> 
> * queryNode与origin dataset的size相差四倍，这样子是正常的嘛？有没有什么办法可以减少内存使用呢？如果一直按照四倍线性增长的话，可能会超出我们对memory的预期，而且似乎和yihua前面讲的2~3倍有一些比较大的出入；
> * python sdk 2.2.3似乎还没有发布，想问一下有一个大概的预估发布时间嘛？如果没有的话，请问你们有公开getLoadStatus相关的Restful API给用户嘛？
> * 有另一个关于并发的问题：假设client 1 release了一个collection，并且开始建index，此时client 2要进行查询，load collection，这时候client 2的操作会失败嘛？如果会的话，client 2有什么办法可以知道现在collection有正在进行的建index的操作嘛（比如有暴露接口可以查询正在进行中的任务，或者是检查是否满足load collection的条件嘛）？
> 
> 提前谢谢您的解答！


python sdk 就用2.2.2版本就可以了，sdk和server的版本号只有前两个数字严格对应，最后一个数字可以不一样。
get_load_state()在pymilvus 2.2.2里其实是有的：https://github.com/milvus-io/pymilvus/blob/bcc334321314d69df32ebcf4e91b2e9051aa9691/pymilvus/orm/utility.py#L207

如果server先接收到client1的release请求，然后接收到client2的load请求，那就先release再load。load的过程理论上不受index的影响，如果有的segment没建索引，就会把原始向量加载上来。index的任务通过pymilvus的utility.index_building_progress()可以知道大概进度。
如果server先接收到client2的load请求，然后接收到client1的release请求，那就会影响client2，client2想search时就会报错collection not loaded，这时可以通过get_load_state()来知道状态是unloaded，那就再调用一次load
